### DISCUSSION TITLE
Autoexec.bat file

### DISCUSSION THREAD

**Discussion by NsinghP** [arendst/Tasmota#11691] [Q&A] [answered]
Hello, 

I am planning on using autoexec.bat file to execute “devicename’ command. However, I want .bat file to be executed only after the device boots first time after factory reset. During normal reboots I want the command ignored. Is it possible to structure the .bat file in such a way? 

Regards,
Navi

**Comment by sfromis** [comment, marked as answer]
No conditionals in autoexec.bat. If you want conditional commands (and not running ESP32 with other options like Berry), you could invoke a rule to check for a suitable already-configured condition, and execute commands depending on that. That rule could also be created from autoexec.bat (meaning that it would be replaced during normal boots).

**Reply by NsinghP** [reply]
Will making an autoexec.bat file, that looks for value “devicename = friendly name 1 value” and if that is true then it executes “Devicename = custom name” command, work? What would be the syntax of such a file?

**Reply by sfromis** [reply]
No conditionals in autoexec.bat. See https://tasmota.github.io/docs/Rules/ and take a bit of a learning curve. Hint: The Event command can be used from autoexec.bat to invoke a rule.

**Reply by NsinghP** [reply]
will this work:

ON

"DeviceName":"FriendlyName1 value"  

DO

DeviceName CustomName

ENDON

BREAK

Please correct my Syntax.

**Reply by sfromis** [reply]
No. Do read the link I posted, and take that learning curve. You are so far off that it is not merely a question of correcting syntax details. A rule to access the device name would have to first execute a command to produce JSON including the device name (the Status command will do that. From inside a rule it would need to be wrapped in Backlog, like this: `Backlog Status`, and then you can use a JSON-based trigger to select on the current value. Hint: `on Status#DeviceName ....`

**Reply by NsinghP** [reply]
How about this:

ON

Status#DeviceName=FriendlyName1 value

DO

DeviceName CustomName 

ENDON

Break

**Reply by sfromis** [reply]
Are you expecting that the string "FriendlyName1 value" will be automatically replaced? It is not. If you mean to manually use some value, you're getting there, assuming that you also have a rule to execute the Status command. Do lose that "Break" at the end, makes no sense after endon.

**Reply by NsinghP** [reply]
ON

Backlog status#DeviceName=FriendlyName1 value

DO

Backlog DeviceName CustomName 

ENDON

**Reply by sfromis** [reply]
My previous comment still applies, except for the last sentence about "break".

**Reply by NsinghP** [reply]
How about this:

ON

system#boot

DO

Backlog status#DeviceName

ENDON

ON 

Event#Status DeviceName=FriendlyName1 value

Do

Backlog DeviceName CustomName

ENDON

BREAK

**Reply by barbudor** [reply]
Nothing close to reality

```
rule1  on DeviceName#data do var1 %value% endon
rule1 +on Friendlyname1#data do var2 %value% endon
rule1 +on var2#state=%var1% do backlog DeviceName NewDevName ; FriendlyName1 NewFriendlyname endon
DeviceName
FriendlyName1
```

**Reply by barbudor** [reply]
The problem with this way is that you will write the rule to flash again and again at every boot

**Reply by NsinghP** [reply]
oh but wouldn't the flash cycle stop when the device name status is not "FriendlyName1 value"?

**Reply by sfromis** [reply]
The rule command executed in autoexec.bat is what will cause flash write every boot, before the rule itself is even looked at.

**Reply by barbudor** [reply]
autoexec execute the commands
the 3 first lines registers a new rule
and this at every boot
The condition applies to the execution of the commands in the rule (setting device name and so), not the `Rule1` command that will write the rule over and over.

Now, in normal conditions, a unit is not supposed to restart often
You can also use SaveData 0 as the last command in the backlog so the flash will be written no more except on voluntary restarts
But that's not something perfect
I'm using autoexec.bat only on my tasmota32-bluettoth to enter some settings (aliases) that are not saved in flash
I've not really found another real use case without a feature to make line executions conditional

**Reply by NsinghP** [reply]
> rule1  on DeviceName#data do var1 %value% endon
> rule1 +on Friendlyname1#data do var2 %value% endon
> rule1 +on var2#state=%var1% do backlog DeviceName NewDevName ; FriendlyName1 NewFriendlyname endon
> DeviceName
> FriendlyName1

let me first make sense of the various variations of rule1 you have listed above, then I will feed the autoexec file to pre 9,3.1.2 device and see what happens.

**Reply by ascillato** [reply]
Tasmota checks if the new setting that you want to be saved is the same as the one already stored in the flash. If it is the same, there isn't any new flash write.

Please, check the following example. My `rule1` was empty:

```lua
20:39:59.589 CMD: rule1 on power1#state do power2 %value% endon
20:39:59.592 SRC: WebConsole from 192.168.1.103
20:39:59.595 CMD: Group 0, Index 1, Command "RULE", Data "on power1#state do power2 %value% endon"
20:39:59.599 RUL: Stored uncompressed, would compress from 39 to 27 (-31%)
20:39:59.603 RSL: stat/banopb/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":39,"Free":472,"Rules":"on power1#state do power2 %value% endon"}}
20:40:00.598 CFG: Saved to flash at F8, Count 87, Bytes 4096
20:40:01.467 RSL: tele/banopb/STATE = {"Time":"2021-04-14T20:40:01","Uptime":"0T03:38:27","UptimeSec":13107,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"WirelessNet","BSSId":"84:1E:A3:92:66:BE","Channel":1,"RSSI":98,"Signal":-51,"LinkCount":1,"Downtime":"0T00:00:05"}}
20:40:01.481 RSL: tele/banopb/SENSOR = {"Time":"2021-04-14T20:40:01","Switch1":"OFF","Switch2":"OFF","ANALOG":{"Temperature":39.5},"ENERGY":{"TotalStartTime":"2021-01-30T19:12:54","Total":4.555,"Yesterday":2.055,"Today":4.555,"Period":0,"Power":[0,0],"ApparentPower":[0,0],"ReactivePower":[0,0],"Factor":[0.00,0.00],"Frequency":60,"Voltage":128,"Current":[0.000,0.000]},"TempUnit":"C"}
20:40:09.934 CMD: rule1 on power1#state do power2 %value% endon
20:40:09.937 SRC: WebConsole from 192.168.1.103
20:40:09.940 CMD: Group 0, Index 1, Command "RULE", Data "on power1#state do power2 %value% endon"
20:40:09.944 RUL: Stored uncompressed, would compress from 39 to 27 (-31%)
20:40:09.948 RSL: stat/banopb/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":39,"Free":472,"Rules":"on power1#state do power2 %value% endon"}}
20:40:11.467 RSL: tele/banopb/STATE = {"Time":"2021-04-14T20:40:11","Uptime":"0T03:38:37","UptimeSec":13117,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"WirelessNet","BSSId":"84:1E:A3:92:66:BE","Channel":1,"RSSI":98,"Signal":-51,"LinkCount":1,"Downtime":"0T00:00:05"}}
20:40:11.477 RSL: tele/banopb/SENSOR = {"Time":"2021-04-14T20:40:11","Switch1":"OFF","Switch2":"OFF","ANALOG":{"Temperature":39.3},"ENERGY":{"TotalStartTime":"2021-01-30T19:12:54","Total":4.555,"Yesterday":2.055,"Today":4.555,"Period":0,"Power":[0,0],"ApparentPower":[0,0],"ReactivePower":[0,0],"Factor":[0.00,0.00],"Frequency":60,"Voltage":128,"Current":[0.000,0.000]},"TempUnit":"C"}
```

So, after the first time I change it, Tasmota shows `20:40:00.598 CFG: Saved to flash at F8, Count 87, Bytes 4096`, but after the second time I put the same rule, there isn't any `CFG: Saved` log

So, that `autoexec.bat` shouldn´t wear the flash because it will save the rules just once in flash. You don't need savedata 0 for this case.

**Reply by barbudor** [reply]
true is the rule is on a single line
in my example the rule is build over 3 lines (which is absolutely not necessary I admit)
but in that case because the rule in the 1st line is not equal to the rule in flash, it will write it.

Now what you re pointing is interesting as it means that finally that rule stuff is totally useless !
And we can just set DeviceName and FriendlyName1 in the autoexec without taking care if it's done every time
👍

**Reply by NsinghP** [reply]
does this mean that all I had to do was enter:
 DeviceName MYHOME in .bat file and then on second line enter FreindlyName1 First Room Switch ?

so the following complicated :) strings I came up with would work too?

ON

system#boot

DO

Backlog status#DeviceName

ENDON

ON

Event#Status DeviceName=FriendlyName1 value

Do

Backlog DeviceName CustomName

ENDON

BREAK

**Reply by ascillato** [reply]
You don't need all those rules. Besides, you have several syntax errors. Do not put rules in different lines. All the parts of a rule MUST be in the same line: on trigger do action endon
Do not use break after endon. A BREAK is already an ENDON. They are mutually exclusive. Using both will make Tasmota to not understand what you are trying to do.

So in short. Just use the commands you need without any rule and you will have your desired behaviour
